src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -100,44 +106,78 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
     	if (cacheService.isContainComponent()) { // Item is in cache
     		cacheItem = cacheService.getCacheOfComponent();
     		if (cacheItem.isExpired()) { // cacheItem has expired
-    			// Update the cache in a separate thread
-    			populateCacheTask(cacheItem);
+    			// Refresh the cacheItem in a separate thread
+    			refreshCacheItemTask(cacheItem);
     		}
-    	} else { // Item is not in cache
-    		// Create a new cacheItem object to be stored in cache
-    		cacheItem = new MessageCacheItem();
-    		fetchMessages(cacheItem, VIPCfg.getInstance().getMsgOriginsQueue().iterator());
-
-			if (!cacheItem.getCachedData().isEmpty()) {
-				cacheService.addCacheOfComponent(cacheItem);
-			} else if (!dto.getLocale().equals(ConstantsKeys.SOURCE) && fallbackLocalesIter!=null && fallbackLocalesIter.hasNext()) {
-    			// If failed to fetch message, use MessageCacheItem of the next fallback locale.
-				MessagesDTO fallbackLocaleDTO = new MessagesDTO(dto.getComponent(), fallbackLocalesIter.next().toLanguageTag(), dto.getProductID(), dto.getVersion());
-				cacheItem = new ComponentService(fallbackLocaleDTO).getMessages(fallbackLocalesIter);
-				if (!cacheItem.getCachedData().isEmpty()) {
-					cacheService.addCacheOfComponent(cacheItem);
-				}
+			// If the cacheItem is for a fallback locale, create and store cacheItem for the requested locale in a separate thread.
+    		if (!LocaleUtility.isSameLocale(cacheItem.getLocale(), this.dto.getLocale())) {
+				this.createCacheItemTask(fallbackLocalesIter);
 			}
+    	} else { // Item is not in cache. Create and store cacheItem for the requested locale
+    		cacheItem = createCacheItem(fallbackLocalesIter);
     	}
     	return cacheItem;
     }
-    
-	private void populateCacheTask(MessageCacheItem cacheItem) {
+
+	/**
+	 * Creates a new MessageCacheItem for the DTO and stores it in cache.
+	 *
+	 * @param fallbackLocalesIter The fallback locale queue to use in case of failure.
+	 *
+	 */
+    private MessageCacheItem createCacheItem(Iterator<Locale> fallbackLocalesIter) {
```

Comment:
It seems here has a recursive call: getMessages(Iterator<Locale> fallbackLocalesIter)  -> createCacheItem(Iterator<Locale> fallbackLocalesIter)  -> new ComponentService(fallbackLocaleDTO).getMessages(fallbackLocalesIter)

In my opinion, the hierarchical relation between methods are not clear.

Reply:
What is not clear about it? Could you be more specific?